File: examples/go/clean-api-go/application/queries/get_sales.go
package queries

import (
	"context"

	bberrors "github.com/FreakJazz/backbone/backbone-go/errors"
	"github.com/FreakJazz/backbone/backbone-go/infrastructure/logging"
	bbex "github.com/FreakJazz/backbone/backbone-go/interfaces/responses"
	"github.com/freakjazz/clean-api-go/domain/entities"
	"github.com/freakjazz/clean-api-go/domain/repositories"
)

type GetSalesQuery struct {
	ProductID string
	Page      int
	PageSize  int
}

// Items is []*entities.Sale rather than []map[string]interface{} — see
// GetProductsResult's doc comment for why.
type GetSalesResult struct {
	Items      []*entities.Sale
	TotalCount int
	Page       int
	PageSize   int
}

type GetSalesQueryHandler struct {
	repo   repositories.ISaleRepository
	logger *logging.EnhancedLogger
}

func NewGetSalesQueryHandler(repo repositories.ISaleRepository) *GetSalesQueryHandler {
	return &GetSalesQueryHandler{
		repo:   repo,
		logger: logging.NewEnhancedLogger("clean-api-go").WithLayer("application").WithComponent("GetSalesQueryHandler").WithMethod("Handle"),
	}
}

func (h *GetSalesQueryHandler) Handle(ctx context.Context, q GetSalesQuery) (*GetSalesResult, *bbex.ErrorResponse) {
	log := h.logger

	sales, total, err := h.repo.FindByProductID(ctx, q.ProductID, q.Page, q.PageSize)
	if err != nil {
		log.ErrorWithCode("FindByProductID failed", bberrors.InfraDBFailure.Int(), map[string]interface{}{"error": err.Error()})
		e := bbex.ErrorResponseBuilder.InternalServerError(err.Error())
		return nil, &e
	}

	return &GetSalesResult{Items: sales, TotalCount: total, Page: q.Page, PageSize: q.PageSize}, nil
}

File: examples/python/clean_api_python/application/queries/get_sales.py
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from domain.repositories.sale_repository import ISaleRepository


@dataclass
class GetSalesQuery:
    product_id: Optional[str] = None
    page: int = 1
    page_size: int = 10


@dataclass
class GetSalesResult:
    items: List[Dict[str, Any]]
    total_count: int
    page: int
    page_size: int


class GetSalesQueryHandler:
    def __init__(self, repo: ISaleRepository) -> None:
        self._repo = repo

    def handle(self, query: GetSalesQuery) -> GetSalesResult:
        sales, total = self._repo.find_by_product_id(query.product_id, query.page, query.page_size)
        return GetSalesResult(
            items=[s.to_dict() for s in sales],
            total_count=total,
            page=query.page,
            page_size=query.page_size,
        )
